cpp/tensorrt_llm/kernels/indexerTopK.cu

```diff
@@ -0,0 +1,320 @@
+{
+    union
+    {
+        __half h;
+        uint16_t u16;
+    } tmp;
+
+    tmp.h = __float2half_rn(x);
+    tmp.u16 = (x < 0.f) ? (~tmp.u16 & 0xffff) : (tmp.u16 | 0x8000);
+    return 511 - (tmp.u16 >> 7);
+}
+
+template <int kNumThreadsPerBlock = 512, int kNumBins = 512, int kTopK = 2048>
+__device__ void topKPerRowJob(float const* logits, int const rowStart, int const rowEnd, int const rowIdx,
+    int* outIndices, int stride0, int stride1)
+{
+    // The number of elements per thread for the final top-k sort.
+    static constexpr int kNumTopKItemsPerThread = kTopK / kNumThreadsPerBlock;
+    // The class to sort the elements during the final top-k sort.
+    using TopKSort = cub::BlockRadixSort<float, kNumThreadsPerBlock, kNumTopKItemsPerThread, int>;
+
+    // The number of slots for the final pass.
+    static constexpr int kNumFinalItems = 3072;
+    // The number of elements per thread for the final sort.
+    static constexpr int kNumFinalItemsPerThread = kNumFinalItems / kNumThreadsPerBlock;
+    // The class to sort the elements during the final pass.
+    using FinalSort = cub::BlockRadixSort<float, kNumThreadsPerBlock, kNumFinalItemsPerThread, int>;
+
+    // The class to compute the inclusive prefix-sum over the histogram.
+    using Scan = cub::BlockScan<int, kNumThreadsPerBlock>;
+
+    // Shared memory to compute the block scan.
+    __shared__ typename Scan::TempStorage smemScan;
+
+    // The structure to store the final items (for the final pass).
+    struct FinalItems
+    {
+        // Shared memory to store the indices for the final pass.
+        int indices[kNumFinalItems];
+        // Shared memory to store the logits for the final pass.
+        float logits[kNumFinalItems];
+    };
+
+    // Shared memory to compute the block sort.
+    __shared__ union
+    {
+        FinalItems items;
+        typename FinalSort::TempStorage finalSort;
+        typename TopKSort::TempStorage topKSort;
+    } smemFinal;
+
+    // Shared memory to store the histogram.
+    __shared__ int smemHistogram[kNumBins];
+    // Shared memory to store the selected indices.
+    __shared__ int smemIndices[kTopK];
+    // Shared memory to store the threshold bin.
+    __shared__ int smemThresholdBinIdx[1];
+    // Shared memory counter to register the candidates for the final phase.
+    __shared__ int smemFinalDstIdx[1];
+
+    // The length of the row.
+    int rowLen = rowEnd - rowStart;
+
+    // Shortcut if the length of the row is smaller than Top-K. Indices are not
+    // sorted by their corresponding logit.
+    if (rowLen <= kTopK)
+    {
+        for (int rowIt = threadIdx.x; rowIt < rowLen; rowIt += kNumThreadsPerBlock)
+        {
+            int idx = rowStart + rowIt;
+            outIndices[rowIdx * kTopK + rowIt] = idx - rowStart;
+        }
+        for (int rowIt = rowLen + threadIdx.x; rowIt < kTopK; rowIt += kNumThreadsPerBlock)
+        {
+            outIndices[rowIdx * kTopK + rowIt] = -1;
```

Comment:
nit: Since the kernel already sets invalid entries to -1, can we safely skip the fill_ op (`topk_indices_buffer[:hidden_states.shape[0]] = -1`) when allocating the top-k buffer (at dsa.py around L958)?